python/pyspark/sql/connect/session.py

```diff
@@ -1041,11 +1045,17 @@ def _start_connect_server(master: str, opts: Dict[str, Any]) -> None:
             init_opts.update(opts)
             opts = init_opts
 
+            token = str(uuid.uuid4())
+
             # Configurations to be overwritten
             overwrite_conf = opts
             overwrite_conf["spark.master"] = master
             overwrite_conf["spark.local.connect"] = "1"
+            # When running a local server, always use an ephemeral port
+            overwrite_conf["spark.connect.grpc.binding.port"] = "0"
+            overwrite_conf["spark.connect.authenticate.token"] = token
             os.environ["SPARK_LOCAL_CONNECT"] = "1"
+            os.environ["SPARK_CONNECT_AUTHENTICATE_TOKEN"] = token
```

Comment:
You're making the same mistake as I did :-). Individual Spark tests does not actually terminate JVM but only stop SparkContext. After that, you create the SparkContext (via SparkSession) to Spark connect server here, and it creates a new token. Then the previously set `SPARK_CONNECT_AUTHENTICATE_TOKEN` will be applied to the Spark Connect server, then you end up with a different token. 

To workaround this problem, you should either create the token once (as in https://github.com/apache/spark/pull/49880), or have a variable to reset on `SparkSession.stop`.

Reply:
Isn't that what `del os.environ["SPARK_CONNECT_AUTHENTICATE_TOKEN"]` is doing in `SparkSession.stop`?

I did hit issues with tests and incorrect tokens, which is why I had to manually set the token for the foreach batch worker and query listener processes that are started. The env vars we're stomping on each other. I don't think it's from the token not clearing after stopping, but due to thread-based parallelism that is sharing the same environment vars. I need to go through and remind myself where if anywhere the env vars are actually still needed with how things are setup. I think there were a few unit tests that try to create a second session to the same remote server, which I don't know if that's really even a real use case

Reply:
Ah, okie. I saw you are deleting it. One simpler way is just to generate `token` once only for this process, and don't clean it up at all. The token will be generated per process in any event, so I think it could make the things easier. `spark.connect.authenticate.token` is a static config anyway.

Reply:
Okay I changed things around a little bit:
- The unit tests set a static token value via the config
- When starting the local session, we check if the env var is already set and use that if it is, otherwise if the config is set we set the env var to that, otherwise generate a new token and set the env var
- Don't set the config for the auto started server, just let the env var set the token on the server